src/execution.jl

```diff
@@ -356,92 +423,51 @@ The output of this function is automatically cached, i.e. you can simply call `c
-            dev = device(ctx)
-            cap = supported_capability(dev)
-            asm, kernel_fn, undefined_fns =
-                compile(:ptx, cap, f, tt; name=name, strict=true, kwargs...)
-
-            # settings to JIT based on Julia's debug setting
-            jit_options = Dict{CUDAdrv.CUjit_option,Any}()
-            if Base.JLOptions().debug_level == 1
-                jit_options[CUDAdrv.JIT_GENERATE_LINE_INFO] = true
-            elseif Base.JLOptions().debug_level >= 2
-                jit_options[CUDAdrv.JIT_GENERATE_DEBUG_INFO] = true
-            end
-
-            # link the CUDA device library
-            image = asm
-            # linking the device runtime library requires use of the CUDA linker,
-            # which in turn switches compilation to device relocatable code (-rdc) mode.
-            #
-            # even if not doing any actual calls that need -rdc (i.e., calls to the runtime
-            # library), this significantly hurts performance, so don't do it unconditionally
-            intrinsic_fns = ["vprintf", "malloc", "free", "__assertfail",
-                            "__nvvm_reflect" #= TODO: should have been optimized away =#]
-            if !isempty(setdiff(undefined_fns, intrinsic_fns))
-                @timeit_debug to "device runtime library" begin
-                    linker = CUDAdrv.CuLink(jit_options)
-                    CUDAdrv.add_file!(linker, libcudadevrt[], CUDAdrv.JIT_INPUT_LIBRARY)
-                    CUDAdrv.add_data!(linker, kernel_fn, asm)
-                    image = CUDAdrv.complete(linker)
-                end
-            end
-
-            # JIT into an executable kernel object
-            mod = CuModule(image, jit_options)
-            fun = CuFunction(mod, kernel_fn)
-            kernel = HostKernel{f,tt}(ctx, mod, fun)
-
-            create_exceptions!(mod)
-
-            compilecache[key] = kernel
-            stop = time_ns()
-            @debug begin
-                ver = version(kernel)
-                mem = memory(kernel)
-                reg = registers(kernel)
-                fn = something(name, nameof(f))
-                """Compiled $fn($(join(tt.parameters, ", "))) to PTX $(ver.ptx) for SM $(ver.binary) in $(round((time_ns() - start) / 1000000; digits=2)) ms.
-                   Kernel uses $reg registers, and $(Base.format_bytes(mem.local)) local, $(Base.format_bytes(mem.shared)) shared, and $(Base.format_bytes(mem.constant)) constant memory."""
-            end
-        end
-
-        return compilecache[key]::HostKernel{f,tt}
-    end
+@generated function cufunction(f::Core.Function, tt::Type=Tuple{}; kwargs...)
+    # generated function that crafts a custom code info to call the actual cufunction impl.
+    # this gives us the flexibility to insert manual back edges for automatic recompilation.
+    tt = tt.parameters[1]
+
+    # get a hold of the method and code info of the kernel function
+    sig = Tuple{f, tt.parameters...}
```

Comment:
```suggestion
    sig = Base.signature_type(f, tt)
```

Reply:
This doesn't work here since `f` is already a typeof(function) (as we're in a generator) and signature_type does something weird then:

```
julia> Base.signature_type(sin, Tuple{})
Tuple{typeof(sin)}

julia> Base.signature_type(typeof(sin), Tuple{})
Tuple{Type{typeof(sin)}}
```